crates/forge/tests/integration/mock_call.rs

```diff
@@ -0,0 +1,569 @@
+
+                let dispatcher = IMockCheckerLibCallDispatcher { contract_address };
+                let thing = dispatcher.get_constant_thing_with_lib_call(mock_checker_class_hash);
+                assert(thing == 13, thing);
+            }
+        "#
+        ),
+        Contract::from_code_path(
+            "MockChecker".to_string(),
+            Path::new("tests/data/contracts/mock_checker.cairo"),
+        )
+        .unwrap(),
+        Contract::from_code_path(
+            "MockCheckerLibCall".to_string(),
+            Path::new("tests/data/contracts/mock_checker_library_call.cairo"),
+        )
+        .unwrap()
+    );
+
+    let result = run(
+        &test.path().unwrap(),
+        &String::from("src"),
+        &test.path().unwrap().join("src/lib.cairo"),
+        &Some(test.linked_libraries()),
+        &Default::default(),
+        &corelib_path(),
+        &test.contracts(&corelib_path()).unwrap(),
+        &Utf8PathBuf::from_path_buf(predeployed_contracts().to_path_buf()).unwrap(),
+    )
+    .unwrap();
+
+    assert_passed!(result);
+}
+
+#[test]
+fn start_mock_call_in_constructor_test_has_no_effect() {
+    let test = test_case!(
+        indoc!(
+            r#"
+        use result::ResultTrait;
+        use array::ArrayTrait;
+        use cheatcodes::{ declare, PreparedContract, deploy, start_mock_call };
+
+        #[starknet::interface]
+        trait IConstructorMockChecker<TContractState> {
+            fn get_stored_thing(ref self: TContractState) -> felt252;
+            fn get_constant_thing(ref self: TContractState) -> felt252;
+        }
+
+        #[test]
+        fn start_mock_call_in_constructor_test_has_no_effect() {
+            let class_hash = declare('ConstructorMockChecker');
+            let prepared = PreparedContract { class_hash: class_hash, constructor_calldata: @ArrayTrait::new() };
+            let contract_address = deploy(prepared).unwrap();
+
+            let dispatcher = IConstructorMockCheckerDispatcher { contract_address };
+
+            let mock_ret_data = array![421];
+            start_mock_call(contract_address, 'get_constant_thing', mock_ret_data);
+
+            let thing = dispatcher.get_stored_thing();
+
+            assert(thing == 13, 'Incorrect thing');
+        }
+    "#
+        ),
+        Contract::from_code_path(
+            "ConstructorMockChecker".to_string(),
+            Path::new("tests/data/contracts/constructor_mock_checker.cairo"),
+        )
+        .unwrap()
+    );
+
+    let result = run(
+        &test.path().unwrap(),
+        &String::from("src"),
+        &test.path().unwrap().join("src/lib.cairo"),
+        &Some(test.linked_libraries()),
+        &Default::default(),
+        &corelib_path(),
+        &test.contracts(&corelib_path()).unwrap(),
+        &Utf8PathBuf::from_path_buf(predeployed_contracts().to_path_buf()).unwrap(),
+    )
+    .unwrap();
+    assert_passed!(result);
+}
+
+#[test]
+fn start_mock_call_with_syscall_has_no_effect() {
```

Comment:
It should have an effect, it doesn't in this case since it is an inner call. Can we change it so it is not an inner call and check if it work then?

Reply:
So, how do you want this to work?

Reply:
The same way it is described here, just with `start_mock_call`
https://github.com/foundry-rs/starknet-foundry/pull/181#discussion_r1273118750
It is important to emit an event after mocking bcs validation of block (I think? not sure what exactly is validated) is done on next library call, so we are sure this way that we increment stuff correctly

Reply:
Changed in https://github.com/foundry-rs/starknet-foundry/pull/385/commits/86479a098856b4448a3a14cb916509df46b000e2